Tag PyObject on TensorImpl per torchdeploy interpreter

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56017 Preserve PyObject even when it goes dead
* **#57985 Tag PyObject on TensorImpl per torchdeploy interpreter**

Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28390242](https://our.internmc.facebook.com/intern/diff/D28390242)

ezyang committed:
Tag PyObject on TensorImpl per torchdeploy interpreter

(desc coming soon)

TODO:
- pack into single atomic word on 64-bit
- tighten up interface on TensorImpl instead of exposing mutable ref

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


(desc coming soon)

TODO:
- pack into single atomic word on 64-bit
- tighten up interface on TensorImpl instead of exposing mutable ref

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


(desc coming soon)

TODO:
- pack into single atomic word on 64-bit
- tighten up interface on TensorImpl instead of exposing mutable ref

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28390242).

ezyang commented:
Now updated to have an actual PyInterpreter object as the tag!

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28390242](https://our.internmc.facebook.com/intern/diff/D28390242)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28390242).

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28390242](https://our.internmc.facebook.com/intern/diff/D28390242)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28390242).

ezyang commented:
Planned testing strategy:
- directly share PyObject between two interpreters, error
- race two threads to try to claim a pyobject
- test interpreter printing (need to actually implement)
- test disarm case (allocate tensor, let it live past interpreter destruction)

wconstab commented on c10/core/TensorImpl.h:
```diff
@@ -175,6 +175,109 @@ struct C10_API AutogradMetaFactoryRegisterer {
+
+  // Disarm this PyInterpreter, making all of its methods noops.
+  // This is not really thread safe, but the hope is that long lived tensors
+  // will die after a synchronization even from the dlclose, typically when
```
can you elaborate on the hope and the synchronization?

ezyang replied:
```
  // Disarm this PyInterpreter, making all of its methods noops.
  // Because the function pointers are raw pointers (not atomics),
  // a disarm() invocation that is concurrent with active destructors
  // is not thread safe and will trigger TSAN.  My hope is that this
  // situations doesn't ever actually happen; tensor destruction should
  // quiesce when a dlclose happens, and any long lived tensors whose
  // destructors would be disarmed here only begin the destruction process
  // on process shutdown (long after the dlclose has occurred).

```

wconstab commented on c10/core/TensorImpl.h:
```diff
@@ -1301,12 +1404,89 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+
+  // Clear the PyObject field for an interpreter, in situations where we
+  // statically know the tensor is tagged with our interpreter.
+  void unchecked_clear_pyobj(impl::PyInterpreter* interpreter) {
```
this is just for implementing tests?

ezyang replied:
Oh, this is for the next PR in this stack.

wconstab commented on c10/core/TensorImpl.h:
```diff
@@ -2002,19 +2182,41 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  // http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2055r0.pdf
+  // However, on x86, it doesn't matter if I use acquire or relaxed on the load
+  // as I get the same assembly in both cases.  So I just use the more
+  // conservative acquire (which will impede compiler optimizations but I don't
```
care) ?

wconstab commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -108,12 +101,21 @@ PyObject * THPVariable_Wrap(Variable var)
+      Py_INCREF(obj);
+      return obj;
+    }
+    // TODO: a better invariant is that if we tagged, we MUST have a valid
```
hmm, so this is the case where the interpreter matches, but pyobj_ may have been null.  how does this happen? is it a case where multiple threads on the same interp are racing to do a wrapping?

ezyang replied:
Nope, it's because this PR predates #56017 which guarantees that we don't deallocate the PyObject before Tensor goes dead. Prior to that PR (aka this PR), pyobj will get nulled out by THPVariable_dealloc when all Python side references to it go dead.

wconstab commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -165,7 +167,7 @@ static int THPVariable_clear(THPVariable *self)
     // https://github.com/pytorch/pytorch/issues/22884 for an example of
     // this actually showing up.
-    set_pyobj(self->cdata, nullptr);
+    tensor.unsafeGetTensorImpl()->unchecked_clear_pyobj(self_interpreter.get());
```
i see.  so, this seems ok.  It preserves the invariant that once on an interp, always on that interp.

ezyang replied:
Yup. Clears pyobj but not interpreter tag

wconstab commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -221,7 +223,7 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
   data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
   auto var = data.set_requires_grad(r.toBool(2));
-  return THPVariable_NewWithVar((PyTypeObject*)cls, std::move(var));
+  return THPVariable_NewWithVar((PyTypeObject*)cls, std::move(var), c10::impl::PyInterpreterStatus::DEFINITELY_UNINITIALIZED);
```
i don't understand tensor well enough to know why this is safe and _pynew below is not.  what is var?

ezyang replied:
var is data; the old code is written in a needlessly oblique way. Data is fresh because it is detached `auto data = r.tensor(1).detach();`.

`_pynew` is not OK because if you pass it an integer address it will return a preexisting Tensor object as is, so it could have come from another interpreter in that case.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -55,49 +82,110 @@ static const char* VOLATILE_WARNING =
-static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
+// have a PyObject* associated with it for any interpreter, and also must not
+// be already tagged by another interpreter.  If unchecked, then we assume that
+// our interpreter has already tagged this Varible, and we can do a direct
```
oops comment is old

ezyang replied:
updated commen tis

```
// Creates a new Python object for a Variable.  The status parameter
// specifies what the interpreter tag status on the object is; for
// example, if you ran check_pyobj, the return optional of this object
// tells you if the tensor was already tagged or not so you can pass
// TAGGED_BY_US or MAYBE_UNINITIALIZED; in other cases, you know where
// var came from and can directly assert that it's DEFINITELY_UNINITIALIZED.
// It's ALWAYS safe (albeit slower) to call this with MAYBE_UNINITIALIZED.
```

ezyang replied:
btw, the last sentence isn't true on this version of the PR, but I'm about to post an updated version where it is true.

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28390242](https://our.internmc.facebook.com/intern/diff/D28390242)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28390242).

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28390242](https://our.internmc.facebook.com/intern/diff/D28390242)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28390242).

ezyang committed:
Update on "Tag PyObject on TensorImpl per torchdeploy interpreter"


Fixes https://github.com/pytorch/pytorch/issues/57756

This PR introduces a new `pyobj_interpreter_` field on TensorImpl which tracks what Python interpreter (if any) owns the TensorImpl. This makes it illegal to bind a TensorImpl from multiple Python interpreters, and means that we can now directly store PyObject pointer on TensorImpl even in the presence of multiple Python interpreters, as is the case in torchdeploy. This is a necessary step for PyObject preservation, which cannot be easily implemented when there are multiple Python interpreters.

Although the PR is not that long, there is a very subtle portion of the implementation devoted to ensuring that the tagging process is thread safe, since multiple threads can concurrently try to tag a PyObject. Check Note [Python interpreter tag] and Note [Memory ordering on Python interpreter tag] for detailed discussion of how this is handled. You will have to check this code carefully in code review; I did not torture test the multithreaded paths in any meaningful way.

In a follow up PR, I will pack the interpreter and PyObject fields into single atomic word on 64-bit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28390242](https://our.internmc.facebook.com/intern/diff/D28390242)

[ghstack-poisoned]
